File: packages/google-dai/android/src/main/java/com/bitmovin/player/reactnative/googledai/GoogleDaiModule.kt
package com.bitmovin.player.reactnative.googledai

import android.view.ViewGroup
import com.bitmovin.player.api.Player
import com.bitmovin.player.api.advertising.AdvertisingApi
import com.bitmovin.player.api.event.PlayerEvent
import com.bitmovin.player.integration.googledai.api.GoogleDai
import com.bitmovin.player.integration.googledai.api.GoogleDaiSourceConfig
import com.bitmovin.player.integration.googledai.api.GoogleDaiSourceType
import com.bitmovin.player.integration.googledai.api.googleDai
import com.bitmovin.player.reactnative.NativeId
import com.bitmovin.player.reactnative.PlayerAccessor
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.functions.Queues
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.concurrent.ConcurrentHashMap

private const val GOOGLE_DAI_ARTIFACT = "com.bitmovin.player.integration:google-dai:0.1.0-alpha.1"

class GoogleDaiModule : Module() {
    private val googleDaiAdapters = ConcurrentHashMap<NativeId, GoogleDai>()
    private val googleDaiPlayers = ConcurrentHashMap<NativeId, NativeId>()
    private val playerOwners = ConcurrentHashMap<NativeId, NativeId>()
    private val destroyListeners = ConcurrentHashMap<NativeId, DestroyListener>()

    override fun definition() = ModuleDefinition {
        Name("GoogleDaiModule")

        OnDestroy {
            googleDaiAdapters.keys.toList().forEach(::destroyAdapter)
        }

        AsyncFunction("initialize") { googleDaiId: NativeId, playerId: NativeId ->
            initialize(googleDaiId, playerId)
        }.runOnQueue(Queues.MAIN)

        AsyncFunction("load") { googleDaiId: NativeId, sourceConfig: Map<String, Any?> ->
            val validatedGoogleDaiId = googleDaiId.nonEmptyNativeId("googleDaiId")
            val googleDai = googleDaiAdapters[validatedGoogleDaiId]
                ?: throw GoogleDaiException.UnknownAdapter(validatedGoogleDaiId)
            val nativeSourceConfig = sourceConfig.toGoogleDaiSourceConfig()
            try {
                ensureNativeCompatibility()
                googleDai.load(nativeSourceConfig)
            } catch (error: GoogleDaiException) {
                throw error
            } catch (error: LinkageError) {
                throw GoogleDaiException.NativeCompatibilityFailed(error.message ?: "Unknown error")
            } catch (error: Exception) {
                throw GoogleDaiException.NativeLoadFailed(error.message ?: "Unknown error")
            }
        }.runOnQueue(Queues.MAIN)

        AsyncFunction("destroy") { googleDaiId: NativeId ->
            destroyAdapter(googleDaiId.nonEmptyNativeId("googleDaiId"))
        }.runOnQueue(Queues.MAIN)
    }

    private fun initialize(googleDaiId: NativeId, playerId: NativeId) {
        val validatedGoogleDaiId = googleDaiId.nonEmptyNativeId("googleDaiId")
        val validatedPlayerId = playerId.nonEmptyNativeId("playerId")

        googleDaiAdapters[validatedGoogleDaiId]?.let {
            if (googleDaiPlayers[validatedGoogleDaiId] == validatedPlayerId) {
                return
            }
            throw GoogleDaiException.DuplicateAdapterId(validatedGoogleDaiId)
        }

        playerOwners[validatedPlayerId]?.let { existingGoogleDaiId ->
            if (googleDaiAdapters.containsKey(existingGoogleDaiId)) {
                throw GoogleDaiException.DuplicatePlayer(validatedPlayerId)
            }
            playerOwners.remove(validatedPlayerId, existingGoogleDaiId)
        }

        val player = PlayerAccessor.get(validatedPlayerId)
            ?: throw GoogleDaiException.PlayerUnavailable(validatedPlayerId)
        try {
            ensureNativeCompatibility()
            googleDaiAdapters[validatedGoogleDaiId] = player.googleDai
            googleDaiPlayers[validatedGoogleDaiId] = validatedPlayerId
            playerOwners[validatedPlayerId] = validatedGoogleDaiId
            observePlayerDestroy(validatedGoogleDaiId, player)
        } catch (error: GoogleDaiException) {
            destroyAdapter(validatedGoogleDaiId)
            throw error
        } catch (error: LinkageError) {
            destroyAdapter(validatedGoogleDaiId)
            throw GoogleDaiException.NativeCompatibilityFailed(error.message ?: "Unknown error")
        } catch (error: Exception) {
            destroyAdapter(validatedGoogleDaiId)
            throw GoogleDaiException.NativeInitializationFailed(error.message ?: "Unknown error")
        }
    }

    private fun observePlayerDestroy(googleDaiId: NativeId, player: Player) {
        val listener: (PlayerEvent.Destroy) -> Unit = { destroyAdapter(googleDaiId) }
        player.on(PlayerEvent.Destroy::class, listener)
        destroyListeners[googleDaiId] = DestroyListener(player, listener)
    }

    private fun destroyAdapter(googleDaiId: NativeId) {
        val googleDai = googleDaiAdapters.remove(googleDaiId)
        val playerId = googleDaiPlayers.remove(googleDaiId)
        if (playerId != null) {
            playerOwners.remove(playerId, googleDaiId)
        }
        destroyListeners.remove(googleDaiId)?.remove()
        googleDai?.destroyIfAvailable()
    }
}

private fun NativeId.nonEmptyNativeId(field: String): NativeId {
    if (isBlank()) {
        throw GoogleDaiException.InvalidNativeId(field)
    }
    return this
}

private fun GoogleDai.destroyIfAvailable() {
    try {
        val destroyMethod = javaClass.methods.firstOrNull { method ->
            method.name == "destroy" && method.parameterTypes.isEmpty()
        } ?: javaClass.declaredMethods.firstOrNull { method ->
            method.name == "destroy" && method.parameterTypes.isEmpty()
        } ?: return
        destroyMethod.isAccessible = true
        destroyMethod.invoke(this)
    } catch (_: Exception) {
        // The alpha Android DAI API has no public destroy contract; keep RN destroy idempotent.
    }
}

private fun ensureNativeCompatibility() {
    val hasRequiredAdvertisingViewGroupGetter = AdvertisingApi::class.java.methods.any { method ->
        method.name == "getViewGroup" &&
            method.parameterTypes.isEmpty() &&
            method.returnType == ViewGroup::class.java
    }
    if (!hasRequiredAdvertisingViewGroupGetter) {
        throw GoogleDaiException.NativeCompatibilityFailed(
            "The pinned $GOOGLE_DAI_ARTIFACT artifact requires AdvertisingApi.getViewGroup(), " +
                "but the aligned Android Player SDK does not expose it. Use a mutually compatible " +
                "Android Player SDK and Google DAI artifact pair",
        )
    }
}

private class DestroyListener(
    private val player: Player,
    private val listener: (PlayerEvent.Destroy) -> Unit,
) {
    fun remove() {
        try {
            player.off(PlayerEvent.Destroy::class, listener)
        } catch (_: Exception) {
            // The player can already be in teardown when this listener runs.
        }
    }
}

private fun Map<String, Any?>.toGoogleDaiSourceConfig(): GoogleDaiSourceConfig {
    val kind = this["kind"] as? String ?: throw GoogleDaiException.InvalidSourceConfig(
        "kind must be 'live'",
    )
    if (kind != "live") {
        throw GoogleDaiException.InvalidSourceConfig("unsupported kind '$kind'")
    }
    return GoogleDaiSourceConfig.Live(
        assetKey = nonEmptyString("assetKey"),
        type = sourceType(),
        apiKey = optionalString("apiKey"),
        networkCode = optionalString("networkCode"),
        adTagParameters = adTagParameters(),
    )
}

private fun Map<String, Any?>.nonEmptyString(key: String): String {
    val value = this[key] as? String ?: throw GoogleDaiException.InvalidSourceConfig(
        "$key must be a non-empty string",
    )
    if (value.isBlank()) {
        throw GoogleDaiException.InvalidSourceConfig("$key must be a non-empty string")
    }
    return value
}

private fun Map<String, Any?>.optionalString(key: String): String? {
    val value = this[key] ?: return null
    return value as? String ?: throw GoogleDaiException.InvalidSourceConfig(
        "$key must be a string when provided",
    )
}

private fun Map<String, Any?>.sourceType(): GoogleDaiSourceType =
    when (this["type"] as? String) {
        "dash" -> GoogleDaiSourceType.Dash
        "hls" -> GoogleDaiSourceType.Hls
        else -> throw GoogleDaiException.InvalidSourceConfig("type must be 'dash' or 'hls'")
    }

private fun Map<String, Any?>.adTagParameters(): Map<String, String> {
    val rawValue = this["adTagParameters"] ?: return emptyMap()
    val rawMap = rawValue as? Map<*, *> ?: throw GoogleDaiException.InvalidSourceConfig(
        "adTagParameters must be an object with string keys and values",
    )
    return rawMap.entries.associate { (key, value) ->
        if (key !is String || value !is String) {
            throw GoogleDaiException.InvalidSourceConfig(
                "adTagParameters must contain only string keys and values",
            )
        }
        key to value
    }
}

sealed class GoogleDaiException(message: String) : CodedException(message) {
    class PlayerUnavailable(playerId: NativeId) : GoogleDaiException(
        "Player '$playerId' is not initialized or has already been destroyed.",
    )

    class DuplicatePlayer(playerId: NativeId) : GoogleDaiException(
        "Player '$playerId' already has an active GoogleDai instance.",
    )

    class DuplicateAdapterId(googleDaiId: NativeId) : GoogleDaiException(
        "GoogleDai '$googleDaiId' is already initialized for another player.",
    )

    class UnknownAdapter(googleDaiId: NativeId) : GoogleDaiException(
        "GoogleDai '$googleDaiId' is not initialized or has already been destroyed.",
    )

    class InvalidSourceConfig(reason: String) : GoogleDaiException(
        "Invalid Google DAI source config: $reason.",
    )

    class InvalidNativeId(field: String) : GoogleDaiException(
        "$field must be a non-empty string.",
    )

    class NativeInitializationFailed(reason: String) : GoogleDaiException(
        "Could not initialize Google DAI: $reason.",
    )

    class NativeCompatibilityFailed(reason: String) : GoogleDaiException(
        "Google DAI native dependency compatibility error: $reason.",
    )

    class NativeLoadFailed(reason: String) : GoogleDaiException(
        "Could not load Google DAI source config: $reason.",
    )
}

File: packages/google-dai/ios/GoogleDaiModule.swift
import BitmovinPlayer
import ExpoModulesCore
import Foundation
import RNBitmovinPlayer

public class GoogleDaiModule: Module {
    private var googleDaiPlayers: [NativeId: NativeId] = [:]
    private var playerOwners: [NativeId: NativeId] = [:]
    private var playerDestroyObservers: [NativeId: PlayerDestroyObserver] = [:]

    public func definition() -> ModuleDefinition {
        Name("GoogleDaiModule")

        OnDestroy { [weak self] in
            self?.playerDestroyObservers.values.forEach { $0.remove() }
            self?.googleDaiPlayers.removeAll()
            self?.playerOwners.removeAll()
            self?.playerDestroyObservers.removeAll()
        }

        AsyncFunction("initialize") { [weak self] (googleDaiId: NativeId, playerId: NativeId) in
            guard let self else { return }
            try self.initialize(
                googleDaiId: try nonEmptyNativeId(googleDaiId, field: "googleDaiId"),
                playerId: try nonEmptyNativeId(playerId, field: "playerId")
            )
        }.runOnQueue(.main)

        AsyncFunction("load") { [weak self] (googleDaiId: NativeId, sourceConfig: [String: Any]) in
            guard let self else { return }
            try self.load(
                googleDaiId: try nonEmptyNativeId(googleDaiId, field: "googleDaiId"),
                sourceConfig: sourceConfig
            )
        }.runOnQueue(.main)

        AsyncFunction("destroy") { [weak self] (googleDaiId: NativeId) in
            guard let self else { return }
            self.destroy(googleDaiId: try nonEmptyNativeId(googleDaiId, field: "googleDaiId"))
        }.runOnQueue(.main)
    }

    private func initialize(googleDaiId: NativeId, playerId: NativeId) throws {
        if let existingPlayerId = googleDaiPlayers[googleDaiId] {
            if existingPlayerId == playerId { return }
            throw googleDaiException(
                "DUPLICATE_GOOGLE_DAI_ID",
                "GoogleDai '\(googleDaiId)' is already initialized for another player."
            )
        }

        if playerOwners[playerId] != nil {
            throw googleDaiException(
                "DUPLICATE_PLAYER",
                "Player '\(playerId)' already has an active GoogleDai instance."
            )
        }

        guard let player = PlayerRegistry.getPlayer(nativeId: playerId) else {
            throw googleDaiException(
                "PLAYER_UNAVAILABLE",
                "Player '\(playerId)' is not initialized or has already been destroyed."
            )
        }

        googleDaiPlayers[googleDaiId] = playerId
        playerOwners[playerId] = googleDaiId
        observePlayerDestroy(googleDaiId: googleDaiId, player: player)
    }

    private func load(googleDaiId: NativeId, sourceConfig: [String: Any]) throws {
        guard googleDaiPlayers[googleDaiId] != nil else {
            throw googleDaiException(
                "UNKNOWN_GOOGLE_DAI_ID",
                "GoogleDai '\(googleDaiId)' is not initialized or has already been destroyed."
            )
        }
        _ = try liveSourceConfig(from: sourceConfig)
        throw googleDaiException(
            "IOS_GOOGLE_DAI_NOT_IMPLEMENTED",
            "The iOS Google DAI native SDK contract still needs SDK owner confirmation."
        )
    }

    private func observePlayerDestroy(googleDaiId: NativeId, player: Player) {
        playerDestroyObservers[googleDaiId]?.remove()
        let observer = PlayerDestroyObserver(player: player) { [weak self] in
            self?.destroy(googleDaiId: googleDaiId)
        }
        player.add(listener: observer)
        playerDestroyObservers[googleDaiId] = observer
    }

    private func destroy(googleDaiId: NativeId) {
        if let playerId = googleDaiPlayers.removeValue(forKey: googleDaiId) {
            playerOwners.removeValue(forKey: playerId)
        }
        playerDestroyObservers.removeValue(forKey: googleDaiId)?.remove()
    }
}

private final class PlayerDestroyObserver: NSObject, PlayerListener {
    private weak var player: Player?
    private let onDestroyPlayer: () -> Void

    init(player: Player, onDestroyPlayer: @escaping () -> Void) {
        self.player = player
        self.onDestroyPlayer = onDestroyPlayer
    }

    func remove() {
        player?.remove(listener: self)
        player = nil
    }

    func onDestroy(_ event: DestroyEvent, player: Player) {
        onDestroyPlayer()
    }
}

private struct LiveSourceConfig {
    let assetKey: String
    let type: String
    let apiKey: String?
    let networkCode: String?
    let adTagParameters: [String: String]
}

private func liveSourceConfig(from sourceConfig: [String: Any]) throws -> LiveSourceConfig {
    guard sourceConfig["kind"] as? String == "live" else {
        throw invalidSourceConfig("kind must be 'live'")
    }
    return LiveSourceConfig(
        assetKey: try nonEmptyString(sourceConfig["assetKey"], field: "assetKey"),
        type: try sourceType(sourceConfig["type"]),
        apiKey: try optionalString(sourceConfig["apiKey"], field: "apiKey"),
        networkCode: try optionalString(sourceConfig["networkCode"], field: "networkCode"),
        adTagParameters: try adTagParameters(sourceConfig["adTagParameters"])
    )
}

private func nonEmptyString(_ value: Any?, field: String) throws -> String {
    guard let string = value as? String,
          !string.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
        throw invalidSourceConfig("\(field) must be a non-empty string")
    }
    return string
}

private func optionalString(_ value: Any?, field: String) throws -> String? {
    guard let value, !(value is NSNull) else { return nil }
    guard let string = value as? String else {
        throw invalidSourceConfig("\(field) must be a string when provided")
    }
    return string
}

private func sourceType(_ value: Any?) throws -> String {
    guard let string = value as? String, ["dash", "hls"].contains(string) else {
        throw invalidSourceConfig("type must be 'dash' or 'hls'")
    }
    return string
}

private func adTagParameters(_ value: Any?) throws -> [String: String] {
    guard let value, !(value is NSNull) else { return [:] }
    guard let rawMap = value as? [String: Any] else {
        throw invalidSourceConfig("adTagParameters must be an object with string values")
    }
    return try rawMap.reduce(into: [:]) { result, entry in
        guard let stringValue = entry.value as? String else {
            throw invalidSourceConfig("adTagParameters must contain only string values")
        }
        result[entry.key] = stringValue
    }
}

private func nonEmptyNativeId(_ value: NativeId, field: String) throws -> NativeId {
    guard !value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
        throw googleDaiException("INVALID_NATIVE_ID", "\(field) must be a non-empty string.")
    }
    return value
}

private func invalidSourceConfig(_ reason: String) -> Exception {
    googleDaiException("INVALID_SOURCE_CONFIG", "Invalid Google DAI source config: \(reason).")
}

private func googleDaiException(_ code: String, _ description: String) -> Exception {
    Exception(name: code, description: description)
}
